compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -2225,16 +2233,77 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                                 spans.push(trait_ref.path.span);
                             }
                             spans.push(self_ty.span);
-                            err.span_note(spans, &msg)
+                            err.span_note(spans, &msg);
+                        }
+                        _ => {
+                            err.note(&msg);
                         }
-                        _ => err.note(&msg),
                     },
-                    _ => err.note(&msg),
+                    [] if self.tcx.trait_is_auto(parent_def_id)
+                        || self.tcx.lang_items().items().contains(&Some(parent_def_id))
+                        || self.tcx.get_diagnostic_name(parent_def_id).is_some() =>
+                    {
+                        // We don't want to mention every `impl` of `Copy` and `Debug` at all.
+                        err.note(&msg);
+                    }
+                    [] => {
+                        err.note(&msg);
+
+                        // Mention *all* the `impl`s, the user might have meant to use one of them.
+                        let mut candidates = vec![];
+                        self.tcx.for_each_impl(parent_def_id, |impl_def_id| {
+                            if let Some(trait_ref) = self.tcx.impl_trait_ref(impl_def_id) {
+                                if !matches!(trait_ref.self_ty().kind(), ty::Param(_)) {
+                                    candidates.push(format!("\n  {}", trait_ref));
+                                }
+                            }
+                        });
+                        candidates.sort();
+                        candidates.dedup();
+                        if !candidates.is_empty() {
+                            let (limit, tail) = if candidates.len() > 9 {
+                                (9, format!("\nand {} more", candidates.len() - 9))
+                            } else {
+                                (candidates.len(), String::new())
+                            };
+                            err.note(&format!(
+                                "the following other implementations where found:{}{}",
+                                candidates[..limit].join(""),
+                                tail
+                            ));
+                        }
+                    }
+                    def_ids => {
+                        err.note(&msg);
+
+                        // Mention *all* the applicable `impl`s, the user might have meant to use one of them.
+                        let mut items: Vec<_> = def_ids
+                            .iter()
+                            .filter_map(|&def_id| self.tcx.impl_trait_ref(def_id))
+                            .map(|trait_ref| format!("\n  {}", trait_ref.self_ty()))
+                            .collect();
+                        items.sort();
+                        items.dedup();
+                        if !items.is_empty() {
```

Comment:
I wonder if we could use a helper function for taking some list, sorting, deduplicating and writing a diagnostic with the first element or the first N elements?

Reply:
Moved both cases in this file to a single closure, but yes, I'm sure there are three or four other impls of this algorithm in the codebase. I'll leave that for a later follow up.